fcrepo-clients/src/main/java/edu/unc/lib/dl/acl/fcrepo4/InheritedAclFactory.java

```diff
@@ -0,0 +1,201 @@
+			if (depth < CONTENT_STARTING_DEPTH) {
+
+				Map<String, Set<String>> objectPrincipalRoles = objectAclFactory.getPrincipalRoles(pathPid);
+
+				// Add this object's principals/roles to the result
+				mergePrincipalRoles(inheritedPrincRoles, objectPrincipalRoles);
+
+			} else {
+				// No roles left, and no more can be added, so further processing is not needed
+				if (inheritedPrincRoles.isEmpty()) {
+					return inheritedPrincRoles;
+				}
+				// Determine the set of patron principals to evaluate going forward
+				if (patronPrincipals == null) {
+					patronPrincipals = getPatronPrincipals(inheritedPrincRoles.keySet());
+				}
+				// No patron principals, so no further changes can occur
+				if (patronPrincipals.size() == 0) {
+					return inheritedPrincRoles;
+				}
+
+				// Evaluate remaining inherited patron roles
+				revokedPatronRoles(pathPid, inheritedPrincRoles, patronPrincipals);
+			}
+		}
+
+		// Units cannot be assigned patron roles, but have an assumed non-inheritable everyone permission
+		// Checking for unit depth + 1 since the counter will be one past the object checked
+		if (depth == UNIT_PATH_DEPTH + 1) {
+			Set<String> roles = new HashSet<>();
+			roles.add(UserRole.canViewOriginals.getPropertyString());
+			inheritedPrincRoles.put("everyone", roles);
+		}
+
+		return inheritedPrincRoles;
+	}
+
+	private void mergePrincipalRoles(Map<String, Set<String>> basePrincRoles, Map<String,
+			Set<String>> addPrincRoles) {
+		if (basePrincRoles.isEmpty()) {
+			basePrincRoles.putAll(addPrincRoles);
+		} else {
+			// Add to inherited data.  If principals overlap, extra roles are added (not overridden)
+			addPrincRoles.forEach((principal, roles) -> {
+				if (basePrincRoles.containsKey(principal)) {
+					basePrincRoles.get(principal).addAll(roles);
+				} else {
+					basePrincRoles.put(principal, roles);
+				}
+			});
+		}
+	}
+
+	private void revokedPatronRoles(PID pid, Map<String, Set<String>> princRoles,
+			Set<String> patronPrincipals) {
+		// Check each remaining patron principal to see if it still has patron access
+		Iterator<String> patronIt = patronPrincipals.iterator();
+		while (patronIt.hasNext()) {
+			String patronPrinc = patronIt.next();
+			Set<String> roles = princRoles.get(patronPrinc);
+			
+			// Patron access revoked for this principal, so remove it from inherited roles
+			if (!objectPermissionEvaluator.hasPatronAccess(pid, patronPrinc, Permission.viewMetadata)) {
+				patronIt.remove();
+				princRoles.remove(patronPrinc);
+			} else if (!roles.contains(UserRole.canViewMetadata.getPropertyString())
+						&& !objectPermissionEvaluator.hasPatronAccess(pid, patronPrinc, Permission.viewOriginal)) {
+				// Has metadata but not original permission, so downgrading
+				roles.clear();
+				roles.add(UserRole.canViewMetadata.getPropertyString());
+			}
+		}
+	}
+
+	@Override
+	public PatronAccess getPatronAccess(PID target) {
+		List<PID> pidPath = getPidPath(target);
+		if (pidPath.size() < 3) {
```

Comment:
The patron access modification property doesn't apply to units and collections.  So I will switch this conditional to use the constants